File: Top200/offer22-GetKthFromEnd.java
package Top200;

class GetKthFromEnd {
    class ListNode {
        int val;
        ListNode next;
        
        ListNode() {}
        
        ListNode(int val) {
            this.val = val;
        }

        ListNode(int val, ListNode next) {
            this.val = val;
            this.next = next;
        }
    }

    public ListNode getKthFromEnd(ListNode head, int k) {
        ListNode dummy = new ListNode(-1);
        ListNode prev = dummy;
        ListNode curr = dummy;
        curr.next = head;

        for (int i = 1; i <= k; i++) {
            curr = curr.next;
        }

        while (curr.next != null) {
            curr = curr.next;
            prev = prev.next;
        }

        return prev.next;
    }
}

File: Top200/offer22-GetKthFromEnd.go
package main

/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
type ListNode struct {
	Val  int
	Next *ListNode
}

func getKthFromEnd(head *ListNode, k int) *ListNode {
	if head == nil || head.Next == nil {
		return head
	}

	// 快慢指针的思想
	fast := head
	for i := 0; i < k; i++ {
		fast = fast.Next
	}

	slow := head
	for fast != nil {
		slow = slow.Next
		fast = fast.Next
	}

	return slow
}
